lib/schema/uuid.js

```diff
@@ -0,0 +1,300 @@
+ *
+ * @param {String} option The option you'd like to set the value for
+ * @param {Any} value value for option
+ * @return {undefined}
+ * @function set
+ * @static
+ * @api public
+ */
+
+SchemaUUID.set = SchemaType.set;
+
+/**
+ * Get/set the function used to cast arbitrary values to UUIDs.
+ *
+ * #### Example:
+ *
+ *     // Make Mongoose refuse to cast UUIDs with 0 length
+ *     const original = mongoose.Schema.Types.UUID.cast();
+ *     mongoose.UUID.cast(v => {
+ *       assert.ok(typeof v === "string" && v.length > 0);
+ *       return original(v);
+ *     });
+ *
+ *     // Or disable casting entirely
+ *     mongoose.UUID.cast(false);
+ *
+ * @param {Function} [caster]
+ * @return {Function}
+ * @function get
+ * @static
+ * @api public
+ */
+
+SchemaUUID.cast = function cast(caster) {
+  if (arguments.length === 0) {
+    return this._cast;
+  }
+  if (caster === false) {
+    caster = this._defaultCaster;
+  }
+  this._cast = caster;
+
+  return this._cast;
+};
+
+/*!
+ * ignore
+ */
+
+SchemaUUID._checkRequired = v => UUID_FORMAT.test(v);
+
+/**
+ * Override the function the required validator uses to check whether a string
+ * passes the `required` check.
+ *
+ * @param {Function} fn
+ * @return {Function}
+ * @function checkRequired
+ * @static
+ * @api public
+ */
+
+SchemaUUID.checkRequired = SchemaType.checkRequired;
+
+/**
+ * Check if the given value satisfies a required validator.
+ *
+ * @param {Any} value
+ * @return {Boolean}
+ * @api public
+ */
+
+SchemaUUID.prototype.checkRequired = function checkRequired(value) {
+  return UUID_FORMAT.test(value);
+};
+
+/**
+ * Casts to UUID
+ *
+ * @param {Object} value
+ * @param {Object} doc
+ * @param {Boolean} init whether this is an initialization cast
+ * @api private
+ */
+
+SchemaUUID.prototype.cast = function(value, doc, init) {
+  if (SchemaType._isRef(this, value, doc, init)) {
+    if (isBsonType(value, 'UUID')) {
+      return value;
+    }
+
+    return this._castRef(value, doc, init);
+  }
+
+  let castFn;
+  if (typeof this._castFunction === 'function') {
+    castFn = this._castFunction;
+  } else if (typeof this.constructor.cast === 'function') {
+    castFn = this.constructor.cast();
+  } else {
+    castFn = SchemaUUID.cast();
+  }
+
+  try {
+    return castFn(value);
+  } catch (error) {
+    throw new CastError(SchemaUUID.schemaName, value, this.path, error, this);
+  }
+};
+
+/*!
+ * ignore
+ */
+
+function handleSingle(val) {
+  return this.cast(val);
+}
+
+/*!
+ * ignore
+ */
+
+function handleArray(val) {
+  return val.map((m) => {
+    return this.cast(m);
+  });
+}
+
+SchemaUUID.prototype.$conditionalHandlers =
+utils.options(SchemaType.prototype.$conditionalHandlers, {
+  $all: handleArray,
+  $gt: handleSingle,
+  $gte: handleSingle,
+  $in: handleArray,
+  $lt: handleSingle,
+  $lte: handleSingle,
+  $ne: handleSingle,
+  $nin: handleArray,
+  $options: handleSingle,
+  $regex: handleSingle
```

Comment:
UUID shouldn't support `$regex` or `$options` because it looks like UUID is stored as a buffer in MongoDB, and MongoDB doesn't let you do `$regex` on buffers

Reply:
thanks, i just copy-pasted that from the old PR, because i have no clue what this actually does or what works and what does not

Reply:
resolved in 47a0d078b1a5bc77767068683043c2a389333eb4